### ISSUE TITLE
4.0 Roadmap

### ISSUE THREAD

**Issue by KnorpelSenf** [telegraf/telegraf#1202]
# 4.0 Major Release Roadmap 

There are a number of tasks and discussions that we need to address (partially, at least), before releasing 4.0 makes sense.

As a quick recap, the main point of 4.0 is the migration to TypeScript, as tracked by #1074. Also, support for Telegram Bot API 5.0 will arrive.

This issue is meant to serve as a discussion forum for all the points below.

- [x] [ `channelMode`:](https://github.com/telegraf/telegraf/pull/1195#issue-528642015) ~~leave it there or~~ throw it out? ~~If leave: respect it in type annotations? Else, types are wrong for users of the feature.~~ There will no longer be a channel mode, but other means to achieve the same thing instead. They are yet to be developed.
- [x] `bot.on`: ~~[refactor](https://github.com/telegraf/telegraf/pull/1195#issuecomment-735421884) or~~ leave as-is. <ins>A new method can be introduced anytime</ins>.
- [x] #1194 needs to be reviewed: ~~**help wanted**. Please checkout the branch `fix-scene-types` and verify it works!~~
- [x] #1195 should be played around with by preferably at least two other people to make sure we didn't screw up: ~~**help wanted**. Please checkout the branch `leverage-unionization` and verify it works!~~
- [x] Example bots have to be updated.
- [x] New example bots should be written that illustrate TypeScript usage.
- [x] #1280
- [x] #1282

**Comment by wojpawlik** [comment]
I managed to shoehorn type guard support into `Composer.optional` (using overloading or `// @ts-expect-error`). I also managed to write a standalone one that does typecheck (how do I name it?). If it lands in 4.0.0 alongside a set of suitable type guards, I don't care what happens to `channelMode`, `Composer.mount`, `Composer::on`, `Context::updateType` and `Context::updateSubTypes`.

Which examples need updating?

Docs improvements are blocked by https://telegraf.js.org/ not pointing to typedoc-generated docs.

In general, docs aren't blocking 4.0.0. Improved docs can be released even in a patch release.

**Comment by KnorpelSenf** [comment]
I'm intrigued—can you share the code for `Composer.optional`?

A standalone version of what exactly?

Generally, I believe we should go through the examples once again after #1195, in case they need to be adapted. I don't have anything specific in mind yet.

I agree about the docs. I removed them from this list so that we can stay focused on the release. Doesn't mean we don't need them eventually.

**Comment by KnorpelSenf** [comment]
Also, I have a suggestion. It is based on https://github.com/telegraf/telegraf/pull/1195#issuecomment-735421884.

First of all: we throw out channelMode. It is a bit odd that some upfront configuration implicitly changes the behaviour of all composition methods. It is much clearer to have concise but explicit listeners for specific event types. This could work as follows:

```ts
// some middleware, just like before
const mw = ctx => {}

// exactly one update type
bot.sees('message' mw) // all message updates
bot.sees('channelPost', mw) // all channel_post updates  (1)
bot.sees(['poll', 'pollAnswer'], mw) // all poll-related updates

// one update type of a subtype
bot.sees('message:text', mw) // all message updates with text
bot.sees('callbackQuery:data', mw) // all callback_query updates with data
bot.sees('editedMessage:photo', mw) // all edited_message updates of messages containing a photo

// MessageEntities and Users can be filtered even further  (2)
bot.sees('message:entities:url', mw) // all updates with messages that contain a URL
bot.sees('editedChannelPost:captionEntities:italic', mw) // all updates about edits to channel posts that contain italic text in the caption
bot.sees('leftChatMember:is_bot', mw) // all updates about a bot that left the chat

// leaving out the part before the colon makes reasonable assumptions
bot.sees(':forward', mw) // all updates that are forwarded messages (effectively equal to `message:forward` and `channelPost:forward`)
bot.sees([':text', ':caption'], mw) // all updates that contain text or a caption, except edits
bot.sees('::cashtag', mw) // all updates that have a cashtag in either text or caption of either message or channelPost, except edits

// useful shortcuts 
bot.sees('edit', mw) // all updates that are edits
bot.sees('edit:text', mw) // all updates that are edits to text messages
bot.sees(':new_chat_members:me', mw) // all updates about the bot itself joining a chat  (3)
```

All methods are fail-fast. They throw immediately when called with wrong syntax, rather than performing the matching in an unexpected way.

There are a few things we need to discuss.

#### About the Annotations in the Example

Regarding the annotations `(1)`, `(2)`, and `(3)` in the code, I have the following things to say.
1. I suggest to use camelCase rather than snake_case, e.g. `channelPost` and not `channel_post`. It does introduce odd renaming, but it is consistent with JS/TS naming conventions as well as the previous names, i.e. `ctx.channelPost`. We should be clear about this in the docs.
2. This could be done in the 4.1 release and skipped for now.
3. Do we even want this? Could also be postponed until 4.1.

#### Naming
This syntax is not compatible with `bot.on`, e.g. `bot.on('text')` would need to be expressed by `bot.sees(':text')`. I think it is the right thing to introduce a new method rather than changing the old one, as that would most likely break 98 % of bots ever written.

Is there a better name than `sees`? I thought about `for`, `with`, `receives`, and `gets`, but I don't like any of them.

#### Implementation Cost
Even though this syntax looks advanced, it should in fact be relatively simple to implement. I'm willing to look into this (no time frame promised) if we agree that it is a good idea.

I have an idea how to keep up performance, even though updates like `'::url'` do require some searching.

**Comment by wojpawlik** [comment]
Interesting proposal, but let's discuss it after I open mine :) you'll get to see the code then. The signature would be
```ts
guard<TUpdate extends tt.Update>(
  guardFn: (update: tt.Update) => update is TUpdate,
  ...fns: ReadonlyArray<Middleware<TContext & SomeTransform<TUpdate>>
): this
```

I'm also not sure about the name, I'm considering `guard` and `shape`.

**Comment by wojpawlik** [comment]
I'll try moving all scenes-related stuff to `telegraf/scenes`, after the typedoc-generated docs are hosted, so it's easy to compare.

**Comment by wojpawlik** [comment]
I pushed `v4.0.0` tag. Please publish it to npm at earliest convenience, @dotcypress.

Tomorrow I'll look into hosting the new docs. After that's done, I'll work on changelog.

**Comment by wojpawlik** [comment]
Don't publish 4.0.0 at this point. Wait for 4.0.1.

**Comment by wojpawlik** [comment]
Ideally, `npm owner add knorpelsenf` so we don't have to bug you every release. I don't have npm account.
